src/cargo/util/command_prelude.rs

```diff
@@ -212,6 +212,13 @@ pub trait AppExt: Sized {
     fn arg_dry_run(self, dry_run: &'static str) -> Self {
         self._arg(opt("dry-run", dry_run))
     }
+
+    fn arg_future_incompat_report(self) -> Self {
+        self._arg(opt(
+            "future-incompat-report",
+            "Ouputs a future incompatibility report at the end of the build (unstable)",
+        ))
+    }
```

Comment:
This is the flag that I mentioned in https://github.com/rust-lang/cargo/pull/8825#issuecomment-751853641. It provides a way to print the report from a script (e.g. on CI) without needing to extract the generated ID from the human-readable output.

Reply:
I feel like the infrastructure supporting both a flag and a subcommand is a bit larger than expected. If we want to cater to the CI use case we could make the `--id` argument optional to `cargo describe-future-incompat`, and if it's omitted we'd print the ID anyway to cross-reference in CI logs.

Reply:
Doing that just re-introduces the problem of getting the wrong report. By passing `--future-incompat-report`, you're guaranteed to see the correct report (since you specified the flag alongside the original command).

Reply:
The biggest worry of that is concurrent processes like the RLS or deeply-nested invocations in other build systems. For CI, however, Rust projects are likely going to execute one command followed by another, so I don't think the original problem will come back. If anyone's watching CI they can always execute the same command locally if things differ.

Reply:
> If anyone's watching CI they can always execute the same command locally if things differ.

The report might contain lints from platform-specific code (e.g. the `winapi` crate), or the actual command being executed might be deep within various CI config files. I've struggled to reproduce CI failures before due to issues like this.

The vast majority of the code in this PR is either independent of the command/flag code, or exists to support writing the report to disk (for the command). Removing the flag would only eliminate a small amount of code, while re-introducing the very problem that the command was supposed to solve.

Reply:
> If anyone's watching CI they can always execute the same command locally if things differ.

It will be very difficult to tell if the command was overwritten. The user will need to find the ID printed out by `cargo describe-future-incompat`, and then look through the CI log to see if there was a command that printed out a *different* ID, which was executed *after* the command which printed out the *correct* ID. In practice, I can't imagine anyone doing this, so an overwritten report will likely go unnoticed.

Having a flag solves this issue completely. It also makes it easier to integrate into an existing CI setup, as there's no need to add a separate command invocation.

Reply:
Another advantage to the flag - CI may test a crate with different combinations of features, which may in turn enable different combinations of dependencies. I could easily see someone doing something like:

```
cargo test --features A
cargo test --features B
cargo describe-future-incompat
```

without realizing that this will not show a report for `cargo test --features A`. With a flag, it's immediately clear that `cargo test --features B --future-incompat-report` will only show a report for that specific command, and will *not* show a report for any other commands.

Reply:
Personally I don't really understand the CI use case of this feature. These are all just warnings which don't fail a CI run and I don't know many folks that comb successful CI logs to see if there are any extra warnings or things like that. I'm also under the assumption that we can remove a lot of the support in this PR if there's only a subcommand. There seems like there's a good amount of gymnastics shuffling around whether or not warnings are printed, whether it's just future incompat, etc. It seems like what should happen is that we catch any future incompat message from the compiler, don't print it during `cargo build`, and save it to disk at the end of the build with a `--id` that can be used to regurgitate them later.

Again I understand the need for `--id`, I'm the one who brought it up. The main thing I'm worried about is concurrent Cargo invocations, not serial ones with different features/platforms/etc. If someone really cares about CI and is really combing all of their successful logs, then having a `describe-future-incompat` after each `cargo` command doesn't seem too bad.

Reply:
> These are all just warnings which don't fail a CI run and I don't know many folks that comb successful CI logs to see if there are any extra warnings or things like that

My intention was to add a way of doing this in an automated way in a follow-up PR (e.g. allow cargo to fail the build when `--future-incompat-report` is non-empty, or have `describe-future-incompat` set a non-zero exit code when the report is non-empty).

From my perspective, the CI use case is the most valuable part of this feature. The idea of this feature is to give users advance notice of when their crates will break. For a larger crate, a user running it locally will likely not test all of the dependencies (for example, a Linux user will never end up building a `winapi` dependency). What a crate author does with this information is up to them (e.g. an allow-fail CI job, or some kind of notification), but I think it's important that we provide the necessary tools for this to be set up. 

> The main thing I'm worried about is concurrent Cargo invocations

Has this come up in practice before for other commands? The two scenarios I can think of are:

1. Some kind of strange build system wrapper that spawns multiple simultaneous Cargo invocations for the same crate
2. A user having some kind of auto-build running (e.g. in their editor), which ends up running at the same time as an explicit `cargo build`.

For either of these scenarios to actually cause problems, they will need to either compile different sets of upstream dependencies, or set different features for upstream dependencies. Simply building the same upstream dependencies multiple times will result in the same report being generated, so it wouldn't which one actually got displayed.

I'm fine with supporting `--id` to try to prevent this, but I'm concerned that an important usecase (detecting future breakage on other platforms which are tested by CI) is being detrimentally affected by a very obscure one.

Reply:
>  There seems like there's a good amount of gymnastics shuffling around whether or not warnings are printed, whether it's just future incompat, etc.

See https://github.com/rust-lang/cargo/pull/8825#discussion_r564666799. This logic is needed to support correctly generating a report when using cached upstream dependencies, and would not go away if we just had the command.

Reply:
> If someone really cares about CI and is really combing all of their successful logs, then having a describe-future-incompat after each cargo command doesn't seem too bad.

They *can*, but I think people might expect the command to print some kind of summary of all commands, instead of the most recent one. Doing:

```rust
cargo build --features A
cargo build --features B
cargo describe-future-incompat
```

will *appear* to work, but will miss any future-incompat messages that only show with with feature 'A'.

Reply:
Sorry this PR grows ever longer with comments and I feel like the same things are being said. I'm not really sure that you're understanding me. Every time I come back to this I have to boot up on ever-more comments about what is going on here and it feels like I'm reading the same thing every time.

I'm saying we need `--id`. I'm ok with adding `cargo build --future-incompat-report` for now as well. Is there anything else here that needs to be discussed?